homeassistant/components/foscam/camera.py

```diff
@@ -28,26 +60,67 @@
     }
 )
 
+SERVICE_PTZ_SCHEMA = vol.Schema(
+    {
+        vol.Required(ATTR_ENTITY_ID): cv.entity_ids,
+        vol.Required(ATTR_MOVEMENT): vol.In(
+            [
+                DIR_UP,
+                DIR_DOWN,
+                DIR_LEFT,
+                DIR_RIGHT,
+                DIR_TOPLEFT,
+                DIR_TOPRIGHT,
+                DIR_BOTTOMLEFT,
+                DIR_BOTTOMRIGHT,
+            ]
+        ),
+        vol.Optional(ATTR_TRAVELTIME, default=DEFAULT_TRAVELTIME): cv.small_float,
+    }
+)
 
-def setup_platform(hass, config, add_entities, discovery_info=None):
+
+async def async_setup_platform(hass, config, async_add_entities, discovery_info=None):
     """Set up a Foscam IP Camera."""
-    add_entities([FoscamCam(config)])
+
+    async def async_handle_ptz(service):
+        """Handle PTZ service call."""
+        movement = service.data[ATTR_MOVEMENT]
+        travel_time = service.data[ATTR_TRAVELTIME]
+        entity_ids = await async_extract_entity_ids(hass, service)
+
+        if not entity_ids:
+            return
+
+        _LOGGER.debug("Moving '%s' camera(s): %s", movement, entity_ids)
+
+        all_cameras = hass.data[FOSCAM_DATA][ENTITIES]
+        target_cameras = [
+            camera for camera in all_cameras if camera.entity_id in entity_ids
+        ]
+
+        for camera in target_cameras:
+            await camera.async_perform_ptz(movement, travel_time)
+
+    hass.services.async_register(
+        DOMAIN, SERVICE_PTZ, async_handle_ptz, schema=SERVICE_PTZ_SCHEMA
+    )
+
+    async_add_entities([FoscamCam(config)])
 
 
 class FoscamCam(Camera):
     """An implementation of a Foscam IP camera."""
 
     def __init__(self, device_info):
         """Initialize a Foscam camera."""
-        from libpyfoscam import FoscamCamera
-
         super().__init__()
 
-        ip_address = device_info.get(CONF_IP)
-        port = device_info.get(CONF_PORT)
-        self._username = device_info.get(CONF_USERNAME)
-        self._password = device_info.get(CONF_PASSWORD)
-        self._name = device_info.get(CONF_NAME)
+        ip_address = device_info[CONF_IP]
+        port = device_info[CONF_PORT]
+        self._username = device_info[CONF_USERNAME]
+        self._password = device_info[CONF_PASSWORD]
+        self._name = device_info[CONF_NAME]
         self._motion_status = False
 
         self._foscam_session = FoscamCamera(
```

Comment:
Create the `FoscamCamera` instance in `async_setup_platform` and pass it to the entity after checking the rtsp port.